### DISCUSSION TITLE
Jotai's onMount vs useEffect

### DISCUSSION THREAD

**Discussion by tpatalas** [pmndrs/jotai#1858] [General] [answered]
In a previous response to an issue on the Jotai, it was recommended to use useEffect and return null to update the Jotai state. 

https://github.com/pmndrs/jotai/issues/750#issue-1012622916

As it has been two years since that response, is this still the recommended approach? Additionally, is it possible to achieve the same result using Jotai's onMount function instead?

Response example shown below from the previous response:

```ts
// performant pattern with write function
const baseCountAtom = atom(0) // internal atom not for export
const effectsForCountAtom = atom([]) // list of effect functions
const countAtom = atom(
  (get) => get(baseCountAtom),
  (get, set, arg) => {
    set(baseCountAtom, arg)
    get(effectsForCountAtom).forEach((fn) => fn(get, set)) // can be delayed with setTimeout
  }
)

// to illustrate one usage
const CountEffect = () => {
  const setEffectsForCount = useUpdateAtom(effectsForCountAtom)
  useEffect(() => {
    const callback = (get, set) => {
      console.log('count is committed:', get(countAtom))
    }
    setEffectsForCount(prev => [...prev, callback])
    return () => {
      setEffectsForCount(prev => prev.filter(fn => fn !== callback))
    }
  }, [])
  return null
}
```

**Comment by dai-shi** [comment, marked as answer]
Yes, useEffect is still a valid approach when we want to integrate jotai atoms and react states/props/contexts.

onMount is preferable as it limits usage. Internally, onMount is called in useEffect, so it's more or less a syntax sugar.

tbh, I'm not exactly share what I meant by "performant".
In summary, the order of preference is:
1. use writable atoms as much as possible
2. use onMount if that helps
3. use useEffect but this isn't very Jotai way. (states aren't fully represented with atoms.)

**Reply by tpatalas** [reply]
Thank you for the clarification.  Didn't know that onMount is called in useEffect. That clears my usage cases.
